client/remote.go

```diff
@@ -326,121 +327,119 @@ func NewGroup(remotes []Remote) (*Group, error) {
 		return nil, errors.New("attempted to create empty remote group")
 	}
 	g := new(Group)
+
 	for _, r := range remotes {
-		heap.Push(g, &item{Remote: r, latency: &ewmaLatency{}})
+		g.remotes = append(g.remotes, item{Remote: r})
 	}
 
 	return g, nil
 }
 
 // Dial returns a connection with best latency measurement.
 func (g *Group) Dial(c *Client) (conn *Conn, err error) {
-	g.Lock()
-	defer g.Unlock()
-
-	if g.Len() == 0 {
+	g.RLock()
+	if len(g.remotes) == 0 {
 		err = errors.New("remote group empty")
-		return
+		return nil, err
+	}
+	// n is the number of trials.
+	// Because of potential expensive fresh tls dial operation,
+	// we limit total dial candidates to a small number.
+	// Also it solves a subtle problem of test 'localhost'
+	// server discovery due to dual ipv6/ipv4 ip resolution.
+	n := 3
+	if len(g.remotes) < n {
+		n = len(g.remotes)
 	}
 
-	var i *item
-	var popped []*item
-	for g.Len() > 0 {
-		i = heap.Pop(g).(*item)
-		popped = append(popped, i)
-		conn, err = i.Dial(c)
-		if err == nil {
-			break
+	remotes := make([]item, n)
+	copy(remotes, g.remotes)
+	g.RUnlock()
+
+	defer func() {
+		g.Lock()
+		if time.Since(g.lastPingAll) > 30*time.Minute {
+			g.lastPingAll = time.Now()
+			go g.PingAll(c)
 		}
+		g.Unlock()
 
-		log.Infof("fail to connect to %v:%v", *i, err)
-		i.latency.Reset()
-		i.errorCount++
-	}
+	}()
 
-	for _, f := range popped {
-		heap.Push(g, f)
+	// shuffle remotes for load balancing
+	for i := range remotes {
+		j := rand.Intn(i + 1)
+		if i != j {
+			remotes[i], remotes[j] = remotes[j], remotes[i]
+		}
```

Comment:
If you're trying to do Fisher-Yates, this seems backwards?
Maybe instead of copying and doing an in-place shuffle, you could do [Fisher-Yates inside-out](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_.22inside-out.22_algorithm).